### ISSUE TITLE
useLoaderData returns stale data

### ISSUE THREAD

**Issue by jaschaio** [remix-run/remix#3536]
### What version of Remix are you using?

1.6.0

### Steps to Reproduce

Here is a sample repository: [git@github.com:jaschaio/remix-use-loader-data.git](https://github.com/jaschaio/remix-use-loader-data)

There are three simple routes:

* `/`
* `/login`
* `/logout`

The `index.jsx` `loader` checks if a cookie exists, and otherwise redirects to `/login`:

```jsx
export const loader = ( { request } ) => {

    const { headers } = request;

    if ( ! headers.get( 'cookie' ) )
        return new Response( null, {
            status: 301,
            headers: {
                Location: '/login',
            },
        } );

    return null;

};
```

The `login.jsx` route defines an `action` that simply set's the cookie and redirects back to the homepage:

```jsx
export const action = async ( { request } ) => {

    const formData = await request.formData();

    const { username } = Object.fromEntries( Array.from( formData.entries() ) )

    return new Response( null, {
        status: 301,
        headers: {
            Location: '/',
            'Set-Cookie': `user=${ username }; Max-Age=2592000; Path=/; SameSite=Lax`,
        },
    } );

};
```

And the `logout.jsx` route just defines an `action` that unsets the cookie:

```jsx
export const action = async () => {

    return new Response( null, {
        status: 301,
        headers: {
            Location: '/',
            'Set-Cookie': `user=; Max-Age=0; Path=/; SameSite=Lax`,
        },
    } );

};
```

The index route further has a logout button, that implements a `fetcher`:

```jsx
export default function Index() {

  const fetcher = useFetcher();

  return (
    <>
        <h1>
            Home
        </h1>
        <button type="button" onClick={ () => fetcher.submit( undefined, { method: 'post', action: '/logout' } ) }>Logout</button>
    </>
  );
}
```

The `root.jsx` `loader` method check's if a cookie is set and returns the username from the cookie:

```js
export const loader = ( { request } ) => {

    const { headers } = request;

    return {
        user: ( headers.get( 'cookie' ) ) ? headers.get( 'cookie' ).split( '=' ).pop() : undefined,
    }

};
```

Copy [the sample repository](https://github.com/jaschaio/remix-use-loader-data) and run `npm run dev`. Then use any username and password to login (this is just a mock).

After you have logged in, click the "Logout" button.

It will still display "Hello $username", because the user is still returned by the `useLoaderData()` call. But I can see within the browsers console network tab that the request to `?_data=root` is successfully made and returns no data as expected.

Only after a hard refresh the data returned from `useLoaderData` is actually empty.

### Expected Behavior

I would expect `useLoaderData()` to return the data from the last call to `?_data=root` instead of stale data.

### Actual Behavior

`useLoaderData()` returns stale data after an action is performed via `fetcher.submit`

**Comment by brophdawg11** [comment]
Yeah this sounds like #3302 to me - chained redirects should be ok now

**Comment by jaschaio** [comment]
Oh nice, will give it a try in a bit and report back. I did search extensively for "useLoaderData" but didn't find that issue so thanks a lot for the pointer.

**Comment by kiliman** [comment]
Also I would recommend you use the `redirect` helper function instead of building your own redirect response.
```js
return redirect('/', { headers: { 'set-cookie': ... } })
```

**Comment by machour** [comment]
@jaschaio 1.6.1 is out, this should be easier to test.
Let us know if it solves your problem.

**Comment by jaschaio** [comment]
I tested with `1.6.1` but from the changelog it seems like #3302 wasn't actually included in that release.

But I tested with `v0.0.0-nightly-1e32bfa-20220622` which was the release mentioned in the issue and it works.

@machour I tried `v0.0.0-nightly-a609751-20220617` as well but it didn't, but I guess you just mixed up the release as the other one works.

**Comment by brophdawg11** [comment]
@jaschaio Just FYI this is now out in `1.6.2-pre.3`, and should go out in a `1.6.2` stable tomorrow or so
